### ISSUE TITLE
Can't get kubectl describe cronjobs to work

### ISSUE THREAD

**Issue by Scukerman** [kubernetes/kubernetes#45418]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): there's no information about my question.

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): cronjob

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1", GitCommit:"b0b7a323cc5a4a2019b2e9520c21c7830b7f708e", GitTreeState:"clean", BuildDate:"2017-04-03T20:44:38Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1", GitCommit:"b0b7a323cc5a4a2019b2e9520c21c7830b7f708e", GitTreeState:"clean", BuildDate:"2017-04-03T20:33:27Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: `bare-metal`
- **OS** (e.g. from /etc/os-release): `Ubuntu 16.04.2 LTS`
- **Kernel** (e.g. `uname -a`): `Linux MYHOSTNAME 4.4.0-72-generic #93-Ubuntu SMP Fri Mar 31 14:07:41 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`
- **Install tools**:
- **Others**:

**What happened**:
1) Can't get cronjobs scheduled.
2) `kubectl describe cronjob xxx` doesn't work. kubectl uses wrong api version (it uses `batch/v1` instead of `batch/v2alpha1`)

**What you expected to happen**:
1) Working cronjobs
2) Working describe command

**How to reproduce it** (as minimally and precisely as possible):
install fresh 1.6.1 cluster with kubeadm, edit apiserver manifest and add flag `--runtime-config=batch/v2alpha1=true` as it's described in [docs](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/)

**Anything else we need to know**:
Nope.

**Comment by ncdc** [comment]
@soltysh

**Comment by Scukerman** [comment]
@ncdc @soltysh any news?

**Comment by soltysh** [comment]
@Scukerman sorry for the long delay, it looks like only `kubectl describe` suffers from this problem, looking into it. Can you try `kubectl get cronjob` if it's working for you. I've just tested this on 1.6 branch and I confirm `kubectl describe cronjob` is not working as expected, whereas `kubectl get` works.

**Comment by soltysh** [comment]
I've changed the title and added appropriate labels to reflect the current state.

**Comment by soltysh** [comment]
I've double checked this applies to master as well.

**Comment by Scukerman** [comment]
@soltysh
```
$ kubectl get cronjob hello -o wide
NAME      SCHEDULE      SUSPEND   ACTIVE    LAST-SCHEDULE   CONTAINER(S)   IMAGE(S)   SELECTOR
hello     */1 * * * *   False     0         <none>

```

This cronjob I made on 5th of May and it still hasn't been scheduled.

**Comment by soltysh** [comment]
@Scukerman can you provide me with your controller logs at level 5, you can grep them with `cronjob_controller.go` which should return only the ones from this controller.

**Comment by soltysh** [comment]
Forgot to add, that the scheduling is working properly for me.

**Comment by Scukerman** [comment]
@soltysh ah, maybe that's because of HA-cluster I made by myself?
Now I have 3 masters (3 apiservers, 3 controller-managers, `3 schedulers`) built up with kubeadm.
And I just noticed that cronjob uses some `default-scheduler` and none of them marked as default. Can you help me with that?

**Comment by Scukerman** [comment]
I don't know how to set logging level for controller manager. I tried `--v=5` flag but I'm not sure it worked.

```
$ kubectl logs -n kube-system kube-controller-manager-black1
I0511 08:22:10.271856       1 leaderelection.go:179] attempting to acquire leader lease...
I0511 08:22:10.308730       1 leaderelection.go:248] lock is held by Black3 and has not yet expired
I0511 08:22:10.308752       1 leaderelection.go:185] failed to acquire lease kube-system/kube-controller-manager
...
I0511 08:22:13.762899       1 leaderelection.go:248] lock is held by Black3 and has not yet expired
I0511 08:22:13.762935       1 leaderelection.go:185] failed to acquire lease kube-system/kube-controller-manager
I0511 08:22:18.024276       1 leaderelection.go:248] lock is held by Black3 and has not yet expired
...
```

**Comment by Scukerman** [comment]
@soltysh here it is:
```
$ kubectl logs -n kube-system kube-controller-manager-black3 | grep cronjob_controller
I0511 08:37:44.442355       1 cronjob_controller.go:95] Starting CronJob Manager
I0511 08:37:44.496712       1 cronjob_controller.go:110] Found 1 cronjobs
I0511 08:37:44.498889       1 cronjob_controller.go:118] Found 1 jobs
I0511 08:37:44.498916       1 cronjob_controller.go:121] Found 0 groups
E0511 08:37:44.510942       1 cronjob_controller.go:247] Cannot determine if default/hello needs to be started: Too many missed start time (> 100). Set or decrease .spec.startingDeadlineSeconds or check clock skew.
I0511 08:37:44.510958       1 cronjob_controller.go:251] No unmet start times for default/hello
I0511 08:37:54.513998       1 cronjob_controller.go:110] Found 1 cronjobs
I0511 08:37:54.516065       1 cronjob_controller.go:118] Found 1 jobs
I0511 08:37:54.516092       1 cronjob_controller.go:121] Found 0 groups
E0511 08:37:54.518112       1 cronjob_controller.go:247] Cannot determine if default/hello needs to be started: Too many missed start time (> 100). Set or decrease .spec.startingDeadlineSeconds or check clock skew.
I0511 08:37:54.518136       1 cronjob_controller.go:251] No unmet start times for default/hello
I0511 08:38:04.520814       1 cronjob_controller.go:110] Found 1 cronjobs
I0511 08:38:04.538515       1 cronjob_controller.go:118] Found 1 jobs
I0511 08:38:04.538546       1 cronjob_controller.go:121] Found 0 groups
E0511 08:38:04.540567       1 cronjob_controller.go:247] Cannot determine if default/hello needs to be started: Too many missed start time (> 100). Set or decrease .spec.startingDeadlineSeconds or check clock skew.
I0511 08:38:04.540582       1 cronjob_controller.go:251] No unmet start times for default/hello
I0511 08:38:14.543780       1 cronjob_controller.go:110] Found 1 cronjobs
I0511 08:38:14.545811       1 cronjob_controller.go:118] Found 1 jobs
I0511 08:38:14.545836       1 cronjob_controller.go:121] Found 0 groups
E0511 08:38:14.547429       1 cronjob_controller.go:247] Cannot determine if default/hello needs to be started: Too many missed start time (> 100). Set or decrease .spec.startingDeadlineSeconds or check clock skew.
I0511 08:38:14.547444       1 cronjob_controller.go:251] No unmet start times for default/hello
```

**Comment by soltysh** [comment]
> And I just noticed that cronjob uses some default-scheduler and none of them marked as default. Can you help me with that?

I'll defer to @kubernetes/sig-scheduling-misc for help with that. I don't know much about schedulers, sorry.

**Comment by Scukerman** [comment]
I got it to work when I set `startingDeadlineSeconds: 10` but I think it's wrong.

**Comment by soltysh** [comment]
`startingDeadlineSeconds` is to tell the controller how much time it has to start the job after it's actual schedule passes. I'm guessing the edit triggered the controller to pick the cronjob. I wonder why it wasn't picked sooner, though. If you can provide with a reproducer of your situation that would be nice. I think I have a fix for the describe, already.

**Comment by Scukerman** [comment]
> _the edit triggered the controller to pick the cronjob_

You are right. It did.

P.S. Installing OS and trying to reproduce it with brand new cluster

**Comment by Scukerman** [comment]
LOL. For some reason cronjobs started to work and I don't know how so I can't reproduce it :(

**Comment by mikkiweesenaar** [comment]
I do see this (cannot describe a cronjob) problem and can reproduce it.

- kubectl version:
```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.6", GitCommit:"7fa1c1756d8bc963f1a389f4a6937dc71f08ada2", GitTreeState:"clean", BuildDate:"2017-06-16T20:46:19Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.2", GitCommit:"477efc3cbe6a7effca06bd1452fa356e2201e1ee", GitTreeState:"clean", BuildDate:"2017-04-19T20:22:08Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration: Cluster in AWS, kubectl on macOS
- Kubectl running on macOS Sierra 10.12.5
- Kernel: `Darwin Mikkis-MacBook-Pro.local 16.6.0 Darwin Kernel Version 16.6.0: Fri Apr 14 16:21:16 PDT 2017; root:xnu-3789.60.24~6/RELEASE_X86_64 x86_64`
- Installed tools / others: n/a

Manual testing:
```
[13:43:51] ~: kubectl describe cronjob helloworld
Error from server (NotFound): cronjobs.batch "helloworld" not found

[13:45:54] ~: kubectl create -f scratch.yml
cronjob "helloworld" created

[13:46:15] ~: kubectl describe cronjob helloworld
Error from server (NotFound): the server could not find the requested resource

[13:46:20] ~: kubectl get cronjob
NAME         SCHEDULE    SUSPEND   ACTIVE    LAST-SCHEDULE
helloworld   * * * * *   False     0         <none>
populator    0 3 * * *   False     0         <none>

[13:46:28] ~: kubectl describe cronjob helloworld
Error from server (NotFound): the server could not find the requested resource

[13:46:39] ~: kubectl delete -f scratch.yml
cronjob "helloworld" deleted

[13:46:45] ~: kubectl get cronjob
NAME        SCHEDULE    SUSPEND   ACTIVE    LAST-SCHEDULE
populator   0 3 * * *   False     0         <none>

[13:46:48] ~: kubectl describe cronjob helloworld
Error from server (NotFound): cronjobs.batch "helloworld" not found
```

scratch.yml contains:
```
apiVersion: batch/v2alpha1
kind: CronJob
metadata:
  name: helloworld
spec:
  schedule: "* * * * *"
  successfulJobsHistoryLimit: 2
  failedJobsHistoryLimit: 2
  jobTemplate:
    spec:
      template:
        spec:
          containers:
          - name: helloworld
            image: hello-world
          restartPolicy: OnFailure
```

**Comment by soltysh** [comment]
@Scukerman since now 1.7 is release, which has all those bits fixed can I ask you to upgrade it. It should be sufficient just to upgrade the `kubectl`. There has been several significant changes wrt supporting multi-versioned resources that I doubt would be able to backport to 1.6.

**Comment by Scukerman** [comment]
I have already upgraded but not tested it yet. Thx. Closing this issue due to upgrade.

**Comment by jurgenweber** [comment]
ok, I have this in 1.7.10.... How do I get the cronjob to schedule again?

**Comment by kfox1111** [comment]
I can confirm this issue. kubectl 1.10 with kube cluster 1.10 fails to describe cronjob.
but kubectl 1.8 with kube cluster 1.10 describes as intended.

**Comment by jdalegonzalez** [comment]
The problem clearly still occurs.  What caused this issue to become closed?  kubectl describe cronjobs <job> fails.  Every time.  Is this a re-introduction of a problem?

**Comment by soltysh** [comment]
@jdalegonzalez in which version you're experiencing the issue? Provide me with the necessary details and if it's still the case I'll reopen as needed.

**Comment by soltysh** [comment]
Preferably provide a sample output with `-v=9`.

**Comment by schoukri** [comment]
I am currently seeing the same issue. I am using a Google Kubernetes Engine from macOS Mojave 10.14.

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.7", GitCommit:"0c38c362511b20a098d7cd855f1314dad92c2780", GitTreeState:"clean", BuildDate:"2018-08-20T10:09:03Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.6-gke.2", GitCommit:"384b4eaa132ca9a295fcb3e5dfc74062b257e7df", GitTreeState:"clean", BuildDate:"2018-08-15T00:10:14Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

Listing all cronjobs works just fine:

```
$ kubectl get cronjobs
NAME                      SCHEDULE      SUSPEND   ACTIVE    LAST SCHEDULE   AGE
cronjob-search-indexer    30 * * * *    False     0         54m             15h
```
But trying to describe a specific cronjob does not work:

```
$ kubectl describe cronjob cronjob-search-indexer
Error from server (NotFound): the server could not find the requested resource
```

Here is the output with `-v=9`

I've redacted most of the response body showing the actual spec for the cronjob because it contains confidential information.

```
$ kubectl describe cronjob cronjob-search-indexer -v=9
I1019 13:26:55.340671   34500 loader.go:357] Config loaded from file /Users/sam/.kube/config
I1019 13:26:55.350678   34500 round_trippers.go:387] curl -k -v -XGET  -H "Accept: application/json" -H "User-Agent: kubectl/v1.10.7 (darwin/amd64) kubernetes/0c38c36" https://35.188.28.115/apis/batch/v1beta1/namespaces/default/cronjobs/cronjob-search-indexer
I1019 13:26:55.652344   34500 round_trippers.go:406] GET https://35.188.28.115/apis/batch/v1beta1/namespaces/default/cronjobs/cronjob-search-indexer 200 OK in 301 milliseconds
I1019 13:26:55.652378   34500 round_trippers.go:412] Response Headers:
I1019 13:26:55.652389   34500 round_trippers.go:415]     Audit-Id: f68bfbd2-fb9b-4125-a44c-1886c71b325e
I1019 13:26:55.652400   34500 round_trippers.go:415]     Content-Type: application/json
I1019 13:26:55.652410   34500 round_trippers.go:415]     Content-Length: 3176
I1019 13:26:55.652419   34500 round_trippers.go:415]     Date: Fri, 19 Oct 2018 20:26:55 GMT
I1019 13:26:55.660010   34500 request.go:874] Response Body: {"kind":"CronJob","apiVersion":"batch/v1beta1","metadata":{"name":"cronjob-search-indexer","FULL_CRONJOB_SPEC_REDACTED","status":{"lastScheduleTime":"2018-10-19T19:30:00Z"}}
I1019 13:26:55.661669   34500 round_trippers.go:387] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: kubectl/v1.10.7 (darwin/amd64) kubernetes/0c38c36" https://35.188.28.115/apis/batch/v1/namespaces/default/cronjobs/cronjob-search-indexer
I1019 13:26:55.740424   34500 round_trippers.go:406] GET https://35.188.28.115/apis/batch/v1/namespaces/default/cronjobs/cronjob-search-indexer 404 Not Found in 78 milliseconds
I1019 13:26:55.740460   34500 round_trippers.go:412] Response Headers:
I1019 13:26:55.740471   34500 round_trippers.go:415]     Audit-Id: XXXXXXXXX-XXXX-XXXXXXX
I1019 13:26:55.740482   34500 round_trippers.go:415]     Content-Type: application/json
I1019 13:26:55.740492   34500 round_trippers.go:415]     Content-Length: 174
I1019 13:26:55.740502   34500 round_trippers.go:415]     Date: Fri, 19 Oct 2018 20:26:55 GMT
I1019 13:26:55.743914   34500 request.go:874] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server could not find the requested resource","reason":"NotFound","details":{},"code":404}
I1019 13:26:55.744630   34500 helpers.go:201] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "the server could not find the requested resource",
  "reason": "NotFound",
  "details": {},
  "code": 404
}]
F1019 13:26:55.744664   34500 helpers.go:119] Error from server (NotFound): the server could not find the requested resource
```

**Comment by rmsutton** [comment]
Having the same exact issue here

**Comment by AndresPineros** [comment]
Experiencing this in cluster 1.10.11 with client version 1.10.5

**Comment by AndresPineros** [comment]
Fixed by upgrading client to 1.13.1
